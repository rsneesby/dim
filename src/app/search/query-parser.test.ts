import { parseQuery, lexer, Token, canonicalizeQuery } from './query-parser';

// To update the snapshots, run:
// npx jest --updateSnapshot src/app/search/query-parser.test.ts

// TODO: failed parse - a failed parse should return the parts that didn't fail plus an error code??

// Some of these are contrived, some are from past search parsing issues
const cases = [
  ['is:blue is:haspower -is:maxpower'],
  ['is:blue is:haspower not:maxpower'],
  ['not:maxpower'],
  ['not -not:maxpower'],
  ['not not not:maxpower'],
  ['is:blue is:weapon or is:armor not:maxpower'], // => is:blue and (is:weapon or is:armor) and -is:maxpower
  ['not not:maxpower'],
  ['-is:equipped is:haspower is:incurrentchar'],
  ['-source:garden -source:lastwish sunsetsafter:arrival'],
  ['-is:exotic -is:locked -is:maxpower -is:tagged stat:total:<55'],
  ['(is:weapon is:sniperrifle) or (is:armor modslot:arrival)'],
  ['(is:weapon and is:sniperrifle) or not (is:armor and modslot:arrival)'],
  ['is:weapon and is:sniperrifle or not is:armor and modslot:arrival'], // => (is:weapon and is:sniperrifle) or (-is:armor and modslot:arrival)
  ['is:weapon is:sniperrifle or not is:armor modslot:arrival'], // => is:weapon and (is:sniperrifle or -is:armor) and modslot:arrival
  ['is:weapon is:sniperrifle or is:armor and modslot:arrival'], // => is:weapon and (is:sniperrifle or (-is:armor modslot:arrival))
  ['is:weapon (is:sniperrifle or (is:armor and modslot:arrival))'], // => is:weapon and (is:sniperrifle or (-is:armor modslot:arrival))
  ['-(power:>1000 and -modslot:arrival)'],
  ['( power:>1000 and -modslot:arrival ) '],
  ['- is:exotic - (power:>1000)'],
  ['is:armor2.0'],
  ['not forgotten'], // => -"forgotten"
  ['cluster tracking'], // => "cluster" and "tracking"
  ['name:"Hard Light"'],
  ["name:'Hard Light'"],
  ['name:"Gahlran\'s Right Hand"'],
  ['-witherhoard'],
  ['perk:수집가'],
  ['perk:"수집가"'],
  ['"수집가"'],
  ['수집가'],
  ['is:rocketlauncher -"cluster" -"tracking module"'],
  ['is:rocketlauncher -"cluster" -\'tracking module\''],
  ['is:rocketlauncher (perk:"cluster" or perk:"tracking module")'],
  ['(is:hunter power:>=540) or (is:warlock power:>=560)'],
  ['"grenade launcher reserves"'],
  // These test copy/pasting from somewhere that automatically converts quotes to "smart quotes"
  ['“grenade launcher reserves”'],
  ['‘grenade launcher reserves’'],
];

// Each of these asserts that the first query is the same as the second query once parsed
const equivalentSearches = [
  [
    'is:blue is:weapon or is:armor not:maxpower',
    'is:blue and (is:weapon or is:armor) and -is:maxpower',
  ],
  ['not forgotten', "-'forgotten'"],
  ['cluster tracking', '"cluster" and "tracking"'],
  [
    'is:weapon and is:sniperrifle or not is:armor and modslot:arrival',
    '(is:weapon and is:sniperrifle) or (-is:armor and modslot:arrival)',
  ],
  [
    'is:weapon is:sniperrifle or not is:armor modslot:arrival',
    'is:weapon and (is:sniperrifle or -is:armor) and modslot:arrival',
  ],
  [
    'is:weapon is:sniperrifle or is:armor and modslot:arrival',
    'is:weapon and (is:sniperrifle or (is:armor and modslot:arrival))',
  ],
  [
    'is:rocketlauncher perk:"cluster" or perk:"tracking module"',
    'is:rocketlauncher (perk:"cluster" or perk:"tracking module")',
  ],
  ['is:blue (is:rocketlauncher', 'is:blue is:rocketlauncher'],
];

// Test what we generate as the canonical form of queries. The first is the input,
// the second is the canonical version
const canonicalize = [
  ['is:blue is:haspower not:maxpower', 'is:blue is:haspower -is:maxpower'],
  [
    'is:weapon and is:sniperrifle or not is:armor and modslot:arrival',
    '(-is:armor modslot:arrival) or (is:sniperrifle is:weapon)',
  ],
  [
    'is:rocketlauncher perk:"cluster" or perk:\'tracking module\'',
    'is:rocketlauncher (perk:"tracking module" or perk:cluster)',
  ],
  ['( power:>1000 and -modslot:arrival ) ', '-modslot:arrival power:>1000'],
];

test.each(cases)('parse |%s|', (query) => {
  // Test just the lexer
  const tokens: Token[] = [];
  for (const t of lexer(query)) {
    tokens.push(t);
  }
  expect(tokens).toMatchSnapshot('lexer');

  // Test the full parse tree
  const ast = parseQuery(query);
  expect(ast).toMatchSnapshot('ast');
});

test.each(equivalentSearches)('|%s| is equivalent to |%s|', (firstQuery, secondQuery) => {
  const firstAST = parseQuery(firstQuery);
  const secondAST = parseQuery(secondQuery);
  expect(firstAST).toEqual(secondAST);
});

test.each(canonicalize)('|%s| is canonically |%s|', (query, canonical) => {
  const canonicalized = canonicalizeQuery(parseQuery(query));
  expect(canonicalized).toEqual(canonical);
});
