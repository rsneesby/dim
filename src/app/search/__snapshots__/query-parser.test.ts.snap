// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse |"grenade launcher reserves"|: ast 1`] = `
Object {
  "args": "grenade launcher reserves",
  "op": "filter",
  "type": "keyword",
}
`;

exports[`parse |"grenade launcher reserves"|: lexer 1`] = `
Array [
  Array [
    "filter",
    "keyword",
    "grenade launcher reserves",
  ],
]
`;

exports[`parse |"수집가"|: ast 1`] = `
Object {
  "args": "수집가",
  "op": "filter",
  "type": "keyword",
}
`;

exports[`parse |"수집가"|: lexer 1`] = `
Array [
  Array [
    "filter",
    "keyword",
    "수집가",
  ],
]
`;

exports[`parse |( power:>1000 and -modslot:arrival ) |: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "args": ">1000",
      "op": "filter",
      "type": "power",
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": "arrival",
        "op": "filter",
        "type": "modslot",
      },
    },
  ],
}
`;

exports[`parse |( power:>1000 and -modslot:arrival ) |: lexer 1`] = `
Array [
  Array [
    "(",
  ],
  Array [
    "filter",
    "power",
    ">1000",
  ],
  Array [
    "and",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "modslot",
    "arrival",
  ],
  Array [
    ")",
  ],
]
`;

exports[`parse |(is:hunter power:>=540) or (is:warlock power:>=560)|: ast 1`] = `
Object {
  "op": "or",
  "operands": Array [
    Object {
      "op": "and",
      "operands": Array [
        Object {
          "args": "hunter",
          "op": "filter",
          "type": "is",
        },
        Object {
          "args": ">=540",
          "op": "filter",
          "type": "power",
        },
      ],
    },
    Object {
      "op": "and",
      "operands": Array [
        Object {
          "args": "warlock",
          "op": "filter",
          "type": "is",
        },
        Object {
          "args": ">=560",
          "op": "filter",
          "type": "power",
        },
      ],
    },
  ],
}
`;

exports[`parse |(is:hunter power:>=540) or (is:warlock power:>=560)|: lexer 1`] = `
Array [
  Array [
    "(",
  ],
  Array [
    "filter",
    "is",
    "hunter",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "power",
    ">=540",
  ],
  Array [
    ")",
  ],
  Array [
    "or",
  ],
  Array [
    "(",
  ],
  Array [
    "filter",
    "is",
    "warlock",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "power",
    ">=560",
  ],
  Array [
    ")",
  ],
]
`;

exports[`parse |(is:weapon and is:sniperrifle) or not (is:armor and modslot:arrival)|: ast 1`] = `
Object {
  "op": "or",
  "operands": Array [
    Object {
      "op": "and",
      "operands": Array [
        Object {
          "args": "weapon",
          "op": "filter",
          "type": "is",
        },
        Object {
          "args": "sniperrifle",
          "op": "filter",
          "type": "is",
        },
      ],
    },
    Object {
      "op": "not",
      "operand": Object {
        "op": "and",
        "operands": Array [
          Object {
            "args": "armor",
            "op": "filter",
            "type": "is",
          },
          Object {
            "args": "arrival",
            "op": "filter",
            "type": "modslot",
          },
        ],
      },
    },
  ],
}
`;

exports[`parse |(is:weapon and is:sniperrifle) or not (is:armor and modslot:arrival)|: lexer 1`] = `
Array [
  Array [
    "(",
  ],
  Array [
    "filter",
    "is",
    "weapon",
  ],
  Array [
    "and",
  ],
  Array [
    "filter",
    "is",
    "sniperrifle",
  ],
  Array [
    ")",
  ],
  Array [
    "or",
  ],
  Array [
    "not",
  ],
  Array [
    "(",
  ],
  Array [
    "filter",
    "is",
    "armor",
  ],
  Array [
    "and",
  ],
  Array [
    "filter",
    "modslot",
    "arrival",
  ],
  Array [
    ")",
  ],
]
`;

exports[`parse |(is:weapon is:sniperrifle) or (is:armor modslot:arrival)|: ast 1`] = `
Object {
  "op": "or",
  "operands": Array [
    Object {
      "op": "and",
      "operands": Array [
        Object {
          "args": "weapon",
          "op": "filter",
          "type": "is",
        },
        Object {
          "args": "sniperrifle",
          "op": "filter",
          "type": "is",
        },
      ],
    },
    Object {
      "op": "and",
      "operands": Array [
        Object {
          "args": "armor",
          "op": "filter",
          "type": "is",
        },
        Object {
          "args": "arrival",
          "op": "filter",
          "type": "modslot",
        },
      ],
    },
  ],
}
`;

exports[`parse |(is:weapon is:sniperrifle) or (is:armor modslot:arrival)|: lexer 1`] = `
Array [
  Array [
    "(",
  ],
  Array [
    "filter",
    "is",
    "weapon",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "is",
    "sniperrifle",
  ],
  Array [
    ")",
  ],
  Array [
    "or",
  ],
  Array [
    "(",
  ],
  Array [
    "filter",
    "is",
    "armor",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "modslot",
    "arrival",
  ],
  Array [
    ")",
  ],
]
`;

exports[`parse |- is:exotic - (power:>1000)|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "op": "not",
      "operand": Object {
        "args": "exotic",
        "op": "filter",
        "type": "is",
      },
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": ">1000",
        "op": "filter",
        "type": "power",
      },
    },
  ],
}
`;

exports[`parse |- is:exotic - (power:>1000)|: lexer 1`] = `
Array [
  Array [
    "not",
  ],
  Array [
    "filter",
    "is",
    "exotic",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "not",
  ],
  Array [
    "(",
  ],
  Array [
    "filter",
    "power",
    ">1000",
  ],
  Array [
    ")",
  ],
]
`;

exports[`parse |-(power:>1000 and -modslot:arrival)|: ast 1`] = `
Object {
  "op": "not",
  "operand": Object {
    "op": "and",
    "operands": Array [
      Object {
        "args": ">1000",
        "op": "filter",
        "type": "power",
      },
      Object {
        "op": "not",
        "operand": Object {
          "args": "arrival",
          "op": "filter",
          "type": "modslot",
        },
      },
    ],
  },
}
`;

exports[`parse |-(power:>1000 and -modslot:arrival)|: lexer 1`] = `
Array [
  Array [
    "not",
  ],
  Array [
    "(",
  ],
  Array [
    "filter",
    "power",
    ">1000",
  ],
  Array [
    "and",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "modslot",
    "arrival",
  ],
  Array [
    ")",
  ],
]
`;

exports[`parse |-is:equipped is:haspower is:incurrentchar|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "op": "not",
      "operand": Object {
        "args": "equipped",
        "op": "filter",
        "type": "is",
      },
    },
    Object {
      "args": "haspower",
      "op": "filter",
      "type": "is",
    },
    Object {
      "args": "incurrentchar",
      "op": "filter",
      "type": "is",
    },
  ],
}
`;

exports[`parse |-is:equipped is:haspower is:incurrentchar|: lexer 1`] = `
Array [
  Array [
    "not",
  ],
  Array [
    "filter",
    "is",
    "equipped",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "is",
    "haspower",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "is",
    "incurrentchar",
  ],
]
`;

exports[`parse |-is:exotic -is:locked -is:maxpower -is:tagged stat:total:<55|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "op": "not",
      "operand": Object {
        "args": "exotic",
        "op": "filter",
        "type": "is",
      },
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": "locked",
        "op": "filter",
        "type": "is",
      },
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": "maxpower",
        "op": "filter",
        "type": "is",
      },
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": "tagged",
        "op": "filter",
        "type": "is",
      },
    },
    Object {
      "args": "<55",
      "op": "filter",
      "type": "stat:total",
    },
  ],
}
`;

exports[`parse |-is:exotic -is:locked -is:maxpower -is:tagged stat:total:<55|: lexer 1`] = `
Array [
  Array [
    "not",
  ],
  Array [
    "filter",
    "is",
    "exotic",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "is",
    "locked",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "is",
    "maxpower",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "is",
    "tagged",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "stat:total",
    "<55",
  ],
]
`;

exports[`parse |-source:garden -source:lastwish sunsetsafter:arrival|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "op": "not",
      "operand": Object {
        "args": "garden",
        "op": "filter",
        "type": "source",
      },
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": "lastwish",
        "op": "filter",
        "type": "source",
      },
    },
    Object {
      "args": "arrival",
      "op": "filter",
      "type": "sunsetsafter",
    },
  ],
}
`;

exports[`parse |-source:garden -source:lastwish sunsetsafter:arrival|: lexer 1`] = `
Array [
  Array [
    "not",
  ],
  Array [
    "filter",
    "source",
    "garden",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "source",
    "lastwish",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "sunsetsafter",
    "arrival",
  ],
]
`;

exports[`parse |-witherhoard|: ast 1`] = `
Object {
  "op": "not",
  "operand": Object {
    "args": "witherhoard",
    "op": "filter",
    "type": "keyword",
  },
}
`;

exports[`parse |-witherhoard|: lexer 1`] = `
Array [
  Array [
    "not",
  ],
  Array [
    "filter",
    "keyword",
    "witherhoard",
  ],
]
`;

exports[`parse |cluster tracking|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "args": "cluster",
      "op": "filter",
      "type": "keyword",
    },
    Object {
      "args": "tracking",
      "op": "filter",
      "type": "keyword",
    },
  ],
}
`;

exports[`parse |cluster tracking|: lexer 1`] = `
Array [
  Array [
    "filter",
    "keyword",
    "cluster",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "keyword",
    "tracking",
  ],
]
`;

exports[`parse |is:armor2.0|: ast 1`] = `
Object {
  "args": "armor2.0",
  "op": "filter",
  "type": "is",
}
`;

exports[`parse |is:armor2.0|: lexer 1`] = `
Array [
  Array [
    "filter",
    "is",
    "armor2.0",
  ],
]
`;

exports[`parse |is:blue is:haspower -is:maxpower|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "args": "blue",
      "op": "filter",
      "type": "is",
    },
    Object {
      "args": "haspower",
      "op": "filter",
      "type": "is",
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": "maxpower",
        "op": "filter",
        "type": "is",
      },
    },
  ],
}
`;

exports[`parse |is:blue is:haspower -is:maxpower|: lexer 1`] = `
Array [
  Array [
    "filter",
    "is",
    "blue",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "is",
    "haspower",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "is",
    "maxpower",
  ],
]
`;

exports[`parse |is:blue is:haspower not:maxpower|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "args": "blue",
      "op": "filter",
      "type": "is",
    },
    Object {
      "args": "haspower",
      "op": "filter",
      "type": "is",
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": "maxpower",
        "op": "filter",
        "type": "is",
      },
    },
  ],
}
`;

exports[`parse |is:blue is:haspower not:maxpower|: lexer 1`] = `
Array [
  Array [
    "filter",
    "is",
    "blue",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "is",
    "haspower",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "not",
    "maxpower",
  ],
]
`;

exports[`parse |is:blue is:weapon or is:armor not:maxpower|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "args": "blue",
      "op": "filter",
      "type": "is",
    },
    Object {
      "op": "or",
      "operands": Array [
        Object {
          "args": "weapon",
          "op": "filter",
          "type": "is",
        },
        Object {
          "args": "armor",
          "op": "filter",
          "type": "is",
        },
      ],
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": "maxpower",
        "op": "filter",
        "type": "is",
      },
    },
  ],
}
`;

exports[`parse |is:blue is:weapon or is:armor not:maxpower|: lexer 1`] = `
Array [
  Array [
    "filter",
    "is",
    "blue",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "is",
    "weapon",
  ],
  Array [
    "or",
  ],
  Array [
    "filter",
    "is",
    "armor",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "not",
    "maxpower",
  ],
]
`;

exports[`parse |is:rocketlauncher (perk:"cluster" or perk:"tracking module")|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "args": "rocketlauncher",
      "op": "filter",
      "type": "is",
    },
    Object {
      "op": "or",
      "operands": Array [
        Object {
          "args": "cluster",
          "op": "filter",
          "type": "perk",
        },
        Object {
          "args": "tracking module",
          "op": "filter",
          "type": "perk",
        },
      ],
    },
  ],
}
`;

exports[`parse |is:rocketlauncher (perk:"cluster" or perk:"tracking module")|: lexer 1`] = `
Array [
  Array [
    "filter",
    "is",
    "rocketlauncher",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "(",
  ],
  Array [
    "filter",
    "perk",
    "cluster",
  ],
  Array [
    "or",
  ],
  Array [
    "filter",
    "perk",
    "tracking module",
  ],
  Array [
    ")",
  ],
]
`;

exports[`parse |is:rocketlauncher -"cluster" -"tracking module"|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "args": "rocketlauncher",
      "op": "filter",
      "type": "is",
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": "cluster",
        "op": "filter",
        "type": "keyword",
      },
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": "tracking module",
        "op": "filter",
        "type": "keyword",
      },
    },
  ],
}
`;

exports[`parse |is:rocketlauncher -"cluster" -"tracking module"|: lexer 1`] = `
Array [
  Array [
    "filter",
    "is",
    "rocketlauncher",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "keyword",
    "cluster",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "keyword",
    "tracking module",
  ],
]
`;

exports[`parse |is:rocketlauncher -"cluster" -'tracking module'|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "args": "rocketlauncher",
      "op": "filter",
      "type": "is",
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": "cluster",
        "op": "filter",
        "type": "keyword",
      },
    },
    Object {
      "op": "not",
      "operand": Object {
        "args": "tracking module",
        "op": "filter",
        "type": "keyword",
      },
    },
  ],
}
`;

exports[`parse |is:rocketlauncher -"cluster" -'tracking module'|: lexer 1`] = `
Array [
  Array [
    "filter",
    "is",
    "rocketlauncher",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "keyword",
    "cluster",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "keyword",
    "tracking module",
  ],
]
`;

exports[`parse |is:weapon (is:sniperrifle or (is:armor and modslot:arrival))|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "args": "weapon",
      "op": "filter",
      "type": "is",
    },
    Object {
      "op": "or",
      "operands": Array [
        Object {
          "args": "sniperrifle",
          "op": "filter",
          "type": "is",
        },
        Object {
          "op": "and",
          "operands": Array [
            Object {
              "args": "armor",
              "op": "filter",
              "type": "is",
            },
            Object {
              "args": "arrival",
              "op": "filter",
              "type": "modslot",
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`parse |is:weapon (is:sniperrifle or (is:armor and modslot:arrival))|: lexer 1`] = `
Array [
  Array [
    "filter",
    "is",
    "weapon",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "(",
  ],
  Array [
    "filter",
    "is",
    "sniperrifle",
  ],
  Array [
    "or",
  ],
  Array [
    "(",
  ],
  Array [
    "filter",
    "is",
    "armor",
  ],
  Array [
    "and",
  ],
  Array [
    "filter",
    "modslot",
    "arrival",
  ],
  Array [
    ")",
  ],
  Array [
    ")",
  ],
]
`;

exports[`parse |is:weapon and is:sniperrifle or not is:armor and modslot:arrival|: ast 1`] = `
Object {
  "op": "or",
  "operands": Array [
    Object {
      "op": "and",
      "operands": Array [
        Object {
          "args": "weapon",
          "op": "filter",
          "type": "is",
        },
        Object {
          "args": "sniperrifle",
          "op": "filter",
          "type": "is",
        },
      ],
    },
    Object {
      "op": "and",
      "operands": Array [
        Object {
          "op": "not",
          "operand": Object {
            "args": "armor",
            "op": "filter",
            "type": "is",
          },
        },
        Object {
          "args": "arrival",
          "op": "filter",
          "type": "modslot",
        },
      ],
    },
  ],
}
`;

exports[`parse |is:weapon and is:sniperrifle or not is:armor and modslot:arrival|: lexer 1`] = `
Array [
  Array [
    "filter",
    "is",
    "weapon",
  ],
  Array [
    "and",
  ],
  Array [
    "filter",
    "is",
    "sniperrifle",
  ],
  Array [
    "or",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "is",
    "armor",
  ],
  Array [
    "and",
  ],
  Array [
    "filter",
    "modslot",
    "arrival",
  ],
]
`;

exports[`parse |is:weapon is:sniperrifle or is:armor and modslot:arrival|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "args": "weapon",
      "op": "filter",
      "type": "is",
    },
    Object {
      "op": "or",
      "operands": Array [
        Object {
          "args": "sniperrifle",
          "op": "filter",
          "type": "is",
        },
        Object {
          "op": "and",
          "operands": Array [
            Object {
              "args": "armor",
              "op": "filter",
              "type": "is",
            },
            Object {
              "args": "arrival",
              "op": "filter",
              "type": "modslot",
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`parse |is:weapon is:sniperrifle or is:armor and modslot:arrival|: lexer 1`] = `
Array [
  Array [
    "filter",
    "is",
    "weapon",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "is",
    "sniperrifle",
  ],
  Array [
    "or",
  ],
  Array [
    "filter",
    "is",
    "armor",
  ],
  Array [
    "and",
  ],
  Array [
    "filter",
    "modslot",
    "arrival",
  ],
]
`;

exports[`parse |is:weapon is:sniperrifle or not is:armor modslot:arrival|: ast 1`] = `
Object {
  "op": "and",
  "operands": Array [
    Object {
      "args": "weapon",
      "op": "filter",
      "type": "is",
    },
    Object {
      "op": "or",
      "operands": Array [
        Object {
          "args": "sniperrifle",
          "op": "filter",
          "type": "is",
        },
        Object {
          "op": "not",
          "operand": Object {
            "args": "armor",
            "op": "filter",
            "type": "is",
          },
        },
      ],
    },
    Object {
      "args": "arrival",
      "op": "filter",
      "type": "modslot",
    },
  ],
}
`;

exports[`parse |is:weapon is:sniperrifle or not is:armor modslot:arrival|: lexer 1`] = `
Array [
  Array [
    "filter",
    "is",
    "weapon",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "is",
    "sniperrifle",
  ],
  Array [
    "or",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "is",
    "armor",
  ],
  Array [
    "implicit_and",
  ],
  Array [
    "filter",
    "modslot",
    "arrival",
  ],
]
`;

exports[`parse |name:"Gahlran's Right Hand"|: ast 1`] = `
Object {
  "args": "gahlran's right hand",
  "op": "filter",
  "type": "name",
}
`;

exports[`parse |name:"Gahlran's Right Hand"|: lexer 1`] = `
Array [
  Array [
    "filter",
    "name",
    "gahlran's right hand",
  ],
]
`;

exports[`parse |name:"Hard Light"|: ast 1`] = `
Object {
  "args": "hard light",
  "op": "filter",
  "type": "name",
}
`;

exports[`parse |name:"Hard Light"|: lexer 1`] = `
Array [
  Array [
    "filter",
    "name",
    "hard light",
  ],
]
`;

exports[`parse |name:'Hard Light'|: ast 1`] = `
Object {
  "args": "hard light",
  "op": "filter",
  "type": "name",
}
`;

exports[`parse |name:'Hard Light'|: lexer 1`] = `
Array [
  Array [
    "filter",
    "name",
    "hard light",
  ],
]
`;

exports[`parse |not -not:maxpower|: ast 1`] = `
Object {
  "op": "not",
  "operand": Object {
    "op": "not",
    "operand": Object {
      "op": "not",
      "operand": Object {
        "args": "maxpower",
        "op": "filter",
        "type": "is",
      },
    },
  },
}
`;

exports[`parse |not -not:maxpower|: lexer 1`] = `
Array [
  Array [
    "not",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "not",
    "maxpower",
  ],
]
`;

exports[`parse |not forgotten|: ast 1`] = `
Object {
  "op": "not",
  "operand": Object {
    "args": "forgotten",
    "op": "filter",
    "type": "keyword",
  },
}
`;

exports[`parse |not forgotten|: lexer 1`] = `
Array [
  Array [
    "not",
  ],
  Array [
    "filter",
    "keyword",
    "forgotten",
  ],
]
`;

exports[`parse |not not not:maxpower|: ast 1`] = `
Object {
  "op": "not",
  "operand": Object {
    "op": "not",
    "operand": Object {
      "op": "not",
      "operand": Object {
        "args": "maxpower",
        "op": "filter",
        "type": "is",
      },
    },
  },
}
`;

exports[`parse |not not not:maxpower|: lexer 1`] = `
Array [
  Array [
    "not",
  ],
  Array [
    "not",
  ],
  Array [
    "filter",
    "not",
    "maxpower",
  ],
]
`;

exports[`parse |not not:maxpower|: ast 1`] = `
Object {
  "op": "not",
  "operand": Object {
    "op": "not",
    "operand": Object {
      "args": "maxpower",
      "op": "filter",
      "type": "is",
    },
  },
}
`;

exports[`parse |not not:maxpower|: lexer 1`] = `
Array [
  Array [
    "not",
  ],
  Array [
    "filter",
    "not",
    "maxpower",
  ],
]
`;

exports[`parse |not:maxpower|: ast 1`] = `
Object {
  "op": "not",
  "operand": Object {
    "args": "maxpower",
    "op": "filter",
    "type": "is",
  },
}
`;

exports[`parse |not:maxpower|: lexer 1`] = `
Array [
  Array [
    "filter",
    "not",
    "maxpower",
  ],
]
`;

exports[`parse |perk:"수집가"|: ast 1`] = `
Object {
  "args": "수집가",
  "op": "filter",
  "type": "perk",
}
`;

exports[`parse |perk:"수집가"|: lexer 1`] = `
Array [
  Array [
    "filter",
    "perk",
    "수집가",
  ],
]
`;

exports[`parse |perk:수집가|: ast 1`] = `
Object {
  "args": "수집가",
  "op": "filter",
  "type": "perk",
}
`;

exports[`parse |perk:수집가|: lexer 1`] = `
Array [
  Array [
    "filter",
    "perk",
    "수집가",
  ],
]
`;

exports[`parse |‘grenade launcher reserves’|: ast 1`] = `
Object {
  "args": "grenade launcher reserves",
  "op": "filter",
  "type": "keyword",
}
`;

exports[`parse |‘grenade launcher reserves’|: lexer 1`] = `
Array [
  Array [
    "filter",
    "keyword",
    "grenade launcher reserves",
  ],
]
`;

exports[`parse |“grenade launcher reserves”|: ast 1`] = `
Object {
  "args": "grenade launcher reserves",
  "op": "filter",
  "type": "keyword",
}
`;

exports[`parse |“grenade launcher reserves”|: lexer 1`] = `
Array [
  Array [
    "filter",
    "keyword",
    "grenade launcher reserves",
  ],
]
`;

exports[`parse |수집가|: ast 1`] = `
Object {
  "args": "수집가",
  "op": "filter",
  "type": "keyword",
}
`;

exports[`parse |수집가|: lexer 1`] = `
Array [
  Array [
    "filter",
    "keyword",
    "수집가",
  ],
]
`;
